modules/images/webp-uploads/load.php

```diff
@@ -704,3 +704,159 @@ function webp_uploads_update_rest_attachment( WP_REST_Response $response, WP_Pos
+ * @param int   $attachment_id The ID of the attachment.
+ * @param array $data The current metadata to be stored in the attachment.
+ * @return array The updated metadata of the attachment.
+ */
+function webp_uploads_restore_image( $attachment_id, $data ) {
+	$backup_sources = get_post_meta( $attachment_id, '_wp_attachment_backup_sources', true );
+
+	if ( ! is_array( $backup_sources ) || ! isset( $backup_sources['full-orig'] ) || ! is_array( $backup_sources['full-orig'] ) ) {
+		return $data;
+	}
+
+	// TODO: Handle the case If `IMAGE_EDIT_OVERWRITE` is defined and is truthy remove any edited images if present before replacing the metadata.
+	// See: https://github.com/WordPress/performance/issues/158.
+
+	$data['sources'] = $backup_sources['full-orig'];
+
+	return $data;
+}
+
+/**
+ * Hook fired right after a metadata has been created or updated, this function would look
+ * specifically only for the key: `_wp_attachment_backup_sizes` which is the one used to
+ * store all the backup sizes. This hook is in charge of cleaning up the additional meta
+ * keys stored in the metadata of the attachment and storing the `sources` property in the
+ * previous full size image due this is not a size we need to move the `sources` from the
+ * metadata back into the full size similar as how it's done on the rest of the sizes.
+ *
+ * @since n.e.x.t
+ *
+ * @param int    $meta_id The ID of the meta value stored in the DB.
+ * @param int    $attachment_id The ID of the post used for this meta, in our case the attachment ID.
+ * @param string $meta_name The name of the metadata to be updated.
+ * @param array  $backup_sizes An array with the metadata value in this case the backup sizes.
+ */
+function webp_updated_postmeta( $meta_id, $attachment_id, $meta_name, $backup_sizes ) {
+	// The backup sources array.
+	if ( '_wp_attachment_backup_sizes' !== $meta_name ) {
+		return;
+	}
+
+	$backup_sources = get_post_meta( $attachment_id, '_wp_attachment_backup_sources', true );
+	$backup_sources = is_array( $backup_sources ) ? $backup_sources : array();
+
+	if ( ! isset( $backup_sources['_sources'] ) ) {
+		return;
+	}
+
+	$target = null;
+	foreach ( array_keys( $backup_sizes ) as $size_name ) {
+		// We are only interested in the `full-` sizes.
+		if ( strpos( $size_name, 'full-' ) === false ) {
+			continue;
+		}
+		// If the target already has the sources attributes find the next one.
+		if ( isset( $backup_sources[ $size_name ] ) ) {
+			continue;
+		}
+
+		$target = $size_name;
+		break;
+	}
+
+	if ( null === $target ) {
+		return;
+	}
+
+	$backup_sources[ $target ] = $backup_sources['_sources'];
+	// Remove the temporary used space.
+	unset( $backup_sources['_sources'] );
+	// Store the `sources` property into the full size if present.
+	update_post_meta( $attachment_id, '_wp_attachment_backup_sources', $backup_sources );
+}
+
+add_action( 'added_post_meta', 'webp_updated_postmeta', 10, 4 );
+add_action( 'updated_post_meta', 'webp_updated_postmeta', 10, 4 );
```

Comment:
I still don't understand the need for all of this. We already back up the full image's sources in `webp_uploads_backup_sources()`, so why set that post meta value again here? Why do we need to do _anything_ when `_wp_attachment_backup_sizes` is updated like you're doing here?

Reply:
This is triggered once we know we have sizes on the backup which would allow us to find the right target for the `sources` of the full sizes images. As there can be several versions like:

- `full-orig`
- `full-{hash}`

By hooking after the backup size was created we can determine the correct full target to be allocated for the previous stored `sources` in the `_sources` key of the metadata we are using.